hardware_integration/src/mcp2515_can_interface.cpp

```diff
@@ -0,0 +1,344 @@
+		while (isobus::SystemTiming::get_time_elapsed_ms(start) < 10)
+		{
+			std::uint8_t newMode;
+			if (read_register(MCPRegister::CANSTAT, newMode))
+			{
+				if ((newMode & 0xE0) == static_cast<std::uint8_t>(mode))
+				{
+					retVal = true;
+					break;
+				}
+			}
+		}
+	}
+	return retVal;
+}
+
+bool MCP2515CANInterface::read_frame(isobus::HardwareInterfaceCANFrame &canFrame)
+{
+	constexpr static std::array<RXBRegister, NUM_READ_BUFFERS> RXB = { { { MCPRegister::RXB0CTRL, MCPRegister::RXB0DATA, 0x01 },
+		                                                                   { MCPRegister::RXB1CTRL, MCPRegister::RXB1DATA, 0x02 } } }; ///< The registers of the RX buffers
+	bool retVal = false;
+
+	// Wait for message to be available
+	const struct RXBRegister *rxb;
+
+	while (true)
+	{
+		const std::uint8_t status = get_rx_status();
+
+		// Check if any of the buffers have a message
+		if (status & 0x01)
+		{
+			rxb = &RXB[0];
+			break;
+		}
+		else if (status & 0x02)
+		{
+			rxb = &RXB[1];
+			break;
+		}
+
+		std::this_thread::sleep_for(std::chrono::milliseconds(RECEIVE_MESSAGE_READ_RATE));
+	}
+
+	if (rxb)
+	{
+		std::vector<std::uint8_t> rxData;
+		if (read_register(rxb->ctrl, rxData, 6))
+		{
+			uint32_t id = (rxData[1] << 3) + (rxData[2] >> 5);
+
+			if (0x08 == (rxData[2] & 0x08))
+			{
+				id = (id << 2) + (rxData[2] & 0x03);
+				id = (id << 8) + rxData[3];
+				id = (id << 8) + rxData[4];
+				id |= CAN_EFF_FLAG;
+			}
+
+			uint8_t ctrl = rxData[0];
+			if (ctrl & 0x08)
+			{
+				id |= CAN_RTR_FLAG;
+			}
+
+			uint8_t len = (rxData[5] & 0x0F);
+			if (len <= CAN_MAX_DLEN)
+			{
+				retVal = true;
+
+				if (read_register(rxb->data, rxData, 8) &&
+				    modify_register(MCPRegister::CANINTF, rxb->intf, 0) &&
+				    0 == (id & CAN_ERR_FLAG))
+				{
+					if (0 != (id & CAN_EFF_FLAG))
+					{
+						canFrame.identifier = (id & CAN_EFF_MASK);
+						canFrame.isExtendedFrame = true;
+					}
+					else
+					{
+						canFrame.identifier = (id & CAN_SFF_MASK);
+						canFrame.isExtendedFrame = false;
+					}
+					canFrame.dataLength = len;
+					memset(canFrame.data, 0, sizeof(canFrame.data));
+					memcpy(canFrame.data, rxData.data(), canFrame.dataLength);
+
+					// Timestamp is not supported by MCP2515
+
+					retVal = true;
+				}
+			}
+		}
+	}
+
+	return retVal;
+}
+
+bool MCP2515CANInterface::write_frame(const isobus::HardwareInterfaceCANFrame &canFrame)
+{
+	// The register addresses for the TX buffers
+	constexpr static std::array<TXBRegister, NUM_WRITE_BUFFERS> TXB = { { { MCPRegister::TXB0CTRL, MCPRegister::TXB0SIDH, MCPRegister::TXB0DATA },
+		                                                                    { MCPRegister::TXB1CTRL, MCPRegister::TXB1SIDH, MCPRegister::TXB1DATA },
+		                                                                    { MCPRegister::TXB2CTRL, MCPRegister::TXB2SIDH, MCPRegister::TXB2DATA } } };
+
+	bool retVal = false;
+
+	for (int i = 0; i < NUM_WRITE_BUFFERS; i++)
```

Comment:
Prefer not to use `int` if `NUM_WRITE_BUFFERS` is well defined, should use a fixed-width type